### PR TITLE
Gif upload v1: sprites, costumes

### PR DESCRIPTION
**Pull request by paulkaplan**
### Resolves

_What Github issue does this resolve (please include link)?_

- Resolves https://github.com/LLK/scratch-gui/issues/3541

### Proposed Changes

_Describe what this Pull Request does_

Add support for gif uploading. Handle the different encoding modes for GIFS by drawing each frame to a canvas. 

### Test Coverage

_Please show how you have added tests to cover your changes_

Added integration tests for sprites and costumes to make sure the naming and uploading works correctly.

### CODE REVIEW

**Comment by rschamp** [comment]
Try it out: [gif-upload](https://llk.github.io/scratch-gui/gif-upload)

_I'm a bot, not actually @rschamp!_

**Comment by towerofnix** [comment]
Out of curiosity's sake, what's this "v1"? Are there plans to make a different/improved implementation for gif uploading later? :shipit:

File: src/containers/costume-tab.jsx [hunk 1]
```diff
@@ -205,8 +209,8 @@ class CostumeTab extends React.Component {
     }
     handleCameraBuffer (buffer) {
         const storage = this.props.vm.runtime.storage;
-        const name = this.props.intl.formatMessage(messages.costume, {index: 1});
-        costumeUpload(buffer, 'image/png', name, storage, this.handleNewCostume);
+        const namer = i => this.props.intl.formatMessage(messages.costume, {index: i});
```
Comment (rschamp) [inline src/containers/costume-tab.jsx]: > `namer`

Ok

File: src/containers/target-pane.jsx [hunk 2]
```diff
@@ -147,9 +147,10 @@ class TargetPane extends React.Component {
     }
     handleSpriteUpload (e) {
         const storage = this.props.vm.runtime.storage;
-        const costumeSuffix = this.props.intl.formatMessage(sharedMessages.costume, {index: 1});
         handleFileUpload(e.target, (buffer, fileType, fileName) => {
-            spriteUpload(buffer, fileType, fileName, storage, this.handleNewSprite, costumeSuffix);
+            const pluralizeCostume = i => this.props.intl.formatMessage(
```
Comment (rschamp) [inline src/containers/target-pane.jsx]: > `pluralizeCostume`

Alright fine

File: src/lib/file-uploader.js [hunk 3]
```diff
@@ -194,9 +210,11 @@ const spriteUpload = function (fileData, fileType, spriteName, storage, handleSp
     }
     case 'image/svg+xml':
     case 'image/png':
-    case 'image/jpeg': {
+    case 'image/jpeg':
+    case 'image/gif': {
         // Make a sprite from an image by making it a costume first
-        costumeUpload(fileData, fileType, `${spriteName}-${costumeName}`, storage, (vmCostume => {
+        const costumeNamer = i => `${spriteName}-${costumeSuffixer(i)}`;
```
Comment (rschamp) [inline src/lib/file-uploader.js]: > `costumeNamer`

That makes sense

File: src/lib/file-uploader.js [hunk 4]
```diff
@@ -184,8 +201,7 @@ const soundUpload = function (fileData, fileType, soundName, storage, handleSoun
     handleSound(vmSound);
 };
 
-const spriteUpload = function (fileData, fileType, spriteName, storage, handleSprite, costumeSuffix) {
-    const costumeName = costumeSuffix || 'costume1';
+const spriteUpload = function (fileData, fileType, spriteName, storage, handleSprite, costumeSuffixer) {
```
Comment (rschamp) [inline src/lib/file-uploader.js]: > `costumeSuffixer`

Eh...

File: src/lib/file-uploader.js [hunk 5]
```diff
@@ -86,15 +87,18 @@ const createVMAsset = function (storage, fileName, assetType, dataFormat, data)
...
- * @param {string} costumeName The user-readable name to use for the costume.
+ * @param {string | Function} costumeNameOrNamer String or function producing the user-readable
+ * name to use for the costume. Function will be called with an index in case of gif upload.
  * @param {ScratchStorage} storage The ScratchStorage instance to cache the costume data
  * @param {Function} handleCostume The function to execute on the costume object returned after
  * caching this costume in storage - This function should be responsible for
  * adding the costume to the VM and handling other UI flow that should come after adding the costume
  */
-const costumeUpload = function (fileData, fileType, costumeName, storage, handleCostume) {
+const costumeUpload = function (fileData, fileType, costumeNameOrNamer, storage, handleCostume) {
```
Comment (rschamp) [inline src/lib/file-uploader.js]: > `costumeNameOrNamer`

Ehhhhhh....

**Comment by rschamp** [comment]
@towerofnix The UX for uploading large gifs needs work, so we plan to add notification to the user that stuff is happening while it's happening, and maybe a way to cancel.

File: src/lib/gif-decoder.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        ctx.clearRect(0, 0, canvas.width, canvas.height);
+        ctx.putImageData(imageData, 0, 0);
+
+        const dataUrl = canvas.toDataURL();
+
+        switch (disposal) {
+        case 0:
+        case 1: // Both are "keep previous frame"
+            previousData = ctx.getImageData(0, 0, canvas.width, canvas.height);
+            break;
```
Comment (rschamp) [inline src/lib/gif-decoder.js]: This is the default, so this case could be removed and commented about there.

Comment (paulkaplan) [reply src/lib/gif-decoder.js]: @rschamp fixed in 1d617214b4afb8dbf3f973c8f85d5b03f4be50d7

**Pull request merged**